### PR TITLE
glmmTMB support (#66)

### PR DESCRIPTION
**Pull request by strengejacke**
Address #66

### CODE REVIEW

File: R/estimate_slopes.R [hunk 1]
```diff
@@ -106,9 +119,23 @@ estimate_slopes.merMod <- estimate_slopes.lm
...
-  trends <- tryCatch(emmeans::emtrends(model, levels, var = trend, transform = transform, ...),
-    error = function(e) emmeans::emtrends(model, levels, var = trend, ...)
-  )
+  trends <-
+    tryCatch(
+      {
+        if (component != "conditional") {
+          emmeans::emtrends(model, levels, var = trend, transform = transform, component = "zi", ...)
+        } else {
+          emmeans::emtrends(model, levels, var = trend, transform = transform, ...)
```
Comment (strengejacke) [inline R/estimate_slopes.R:128]: @DominiqueMakowski I don't know why calling `emmeans::emtrends()` directly gives different results as compared to calling it from within `estimate_slopes()`... Both should be the same output:

``` r
require("modelbased")
#> Loading required package: modelbased
require("glmmTMB")
#> Loading required package: glmmTMB
require("emmeans")
#> Loading required package: emmeans

data(Salamanders)
model <- glmmTMB(
  count ~ cover + mined + (1 | site),
  ziformula = ~ cover + mined,
  family = truncated_poisson,
  data = Salamanders
)

estimate_slopes(model, trend = "cover")
#> mined | Coefficient |   SE |  df |         95% CI | Coefficient (std.)
#> ----------------------------------------------------------------------
#> yes   |       -0.20 | 0.06 | 637 | [-0.31, -0.09] |              -0.19
#> no    |       -0.20 | 0.06 | 637 | [-0.31, -0.09] |              -0.19

as.data.frame(emmeans::emtrends(model, "mined", var = "cover", transform = "response"))
#>   mined cover.trend         SE  df  lower.CL   upper.CL
#> 1   yes  -0.2463855 0.07305896 637 -0.389851 -0.1029200
#> 2    no  -0.7646025 0.24201824 637 -1.239852 -0.2893524
```

<sup>Created on 2020-06-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Comment (strengejacke) [reply R/estimate_slopes.R:128]: @mattansb any idea?

Comment (DominiqueMakowski) [reply R/estimate_slopes.R:128]: Sorry haven't had time to look at it, will do!

Comment (mattansb) [reply R/estimate_slopes.R:128]: Without `transform = "response"` they both have the same results:

```R
as.data.frame(emmeans::emtrends(model, "mined", var = "cover"))
#>   mined cover.trend         SE  df   lower.CL    upper.CL
#> 1   yes  -0.1992178 0.05743355 637 -0.3119998 -0.08643586
#> 2    no  -0.1992178 0.05743355 637 -0.3119998 -0.08643586
```

Comment (strengejacke) [reply R/estimate_slopes.R:128]: > Without transform = "response" they both have the same results:

strange thing is, though, that `transform = "response"` by default in `estimate_slopes()`.

Comment (DominiqueMakowski) [reply R/estimate_slopes.R]: sooo it means everything's okay?

Comment (strengejacke) [reply R/estimate_slopes.R]: yes

Comment (strengejacke) [reply R/estimate_slopes.R]: required really an ugly hack (https://github.com/easystats/modelbased/pull/67/commits/bd2c701d228f19f92cb13f4242bd47b04a3e85ae), but.... it seems to work!

**Comment by strengejacke** [comment]
@DominiqueMakowski I'm not sure why you need to set `random = FALSE` here:

https://github.com/easystats/modelbased/blob/5794f17527be2396eafa569bfa45573cbceae44d/R/estimate_smooth.R#L83-L88

however, this means that `visualization_matrix.glmmTMB()` is called with `random = FALSE`, so the matrix does not contain the data from random effects, and thus `predict_wrapper.glmmTMB()` fails. I might need to pass `random` via `...`.

**Comment by strengejacke** [comment]
@DominiqueMakowski ok, you may check everything now... All except `estimate_link()` should work. I can't figure out why estimate_link() fails for my test model?

``` r
require("modelbased")
#> Loading required package: modelbased
require("glmmTMB")
#> Loading required package: glmmTMB

data(Salamanders)
model <- glmmTMB(
  count ~ cover + mined + (1 | site),
  ziformula = ~ cover + mined,
  family = truncated_poisson,
data = Salamanders
)

visualisation_matrix(model, random = FALSE, length = 25, preserve_range = TRUE, reference = insight::get_data(model))
#>          cover mined site
#> 1  -1.59152240   yes   NA
#> 2  -1.44646177   yes   NA
#> 3  -1.30140113   yes   NA
#> 4  -1.15634050   yes   NA
#> 5  -1.01127986   yes   NA
#> 6  -0.86621922   yes   NA
#> 7  -0.72115859   yes   NA
#> 8  -0.57609795   yes   NA
#> 9  -0.43103732   yes   NA
#> 10 -0.28597668   yes   NA
#> 11 -0.14091605   yes   NA
#> 12  0.00414459   yes   NA
#> 13  0.14920523   yes   NA
#> 14  0.29426586   yes   NA
#> 15  0.43932650   yes   NA
#> 16 -0.86621922    no   NA
#> 17 -0.72115859    no   NA
#> 18 -0.57609795    no   NA
#> 19 -0.43103732    no   NA
#> 20 -0.28597668    no   NA
#> 21 -0.14091605    no   NA
#> 22  0.00414459    no   NA
#> 23  0.14920523    no   NA
#> 24  0.29426586    no   NA
#> 25  0.43932650    no   NA
#> 26  0.58438713    no   NA
#> 27  0.72944777    no   NA
#> 28  0.87450840    no   NA
#> 29  1.01956904    no   NA
#> 30  1.16462967    no   NA
#> 31  1.30969031    no   NA
#> 32  1.45475095    no   NA
#> 33  1.59981158    no   NA
#> 34  1.74487222    no   NA
#> 35  1.88993285    no   NA

visualisation_matrix(model, random = TRUE, length = 25, preserve_range = TRUE, reference = insight::get_data(model))
#>          cover mined site
#> 1  -1.59152240   yes   NA
#> 2  -1.44646177   yes   NA
#> 3  -1.30140113   yes   NA
#> 4  -1.15634050   yes   NA
#> 5  -1.01127986   yes   NA
#> 6  -0.86621922   yes   NA
#> 7  -0.72115859   yes   NA
#> 8  -0.57609795   yes   NA
#> 9  -0.43103732   yes   NA
#> 10 -0.28597668   yes   NA
#> 11 -0.14091605   yes   NA
#> 12  0.00414459   yes   NA
#> 13  0.14920523   yes   NA
#> 14  0.29426586   yes   NA
#> 15  0.43932650   yes   NA
#> 16 -0.86621922    no   NA
#> 17 -0.72115859    no   NA
#> 18 -0.57609795    no   NA
#> 19 -0.43103732    no   NA
#> 20 -0.28597668    no   NA
#> 21 -0.14091605    no   NA
#> 22  0.00414459    no   NA
#> 23  0.14920523    no   NA
#> 24  0.29426586    no   NA
#> 25  0.43932650    no   NA
#> 26  0.58438713    no   NA
#> 27  0.72944777    no   NA
#> 28  0.87450840    no   NA
#> 29  1.01956904    no   NA
#> 30  1.16462967    no   NA
#> 31  1.30969031    no   NA
#> 32  1.45475095    no   NA
#> 33  1.59981158    no   NA
#> 34  1.74487222    no   NA
#> 35  1.88993285    no   NA

estimate_slopes(model)
#> No numeric variable was specified for slope estimation. Selecting `trend = "cover"`.
#> mined | Coefficient |   SE |  df |         95% CI | Coefficient (std.)
#> ----------------------------------------------------------------------
#> yes   |       -0.25 | 0.07 | 637 | [-0.39, -0.10] |              -0.24
#> no    |       -0.76 | 0.24 | 637 | [-1.24, -0.29] |              -0.75

estimate_means(model)
#> mined | rate |   SE |       95% CI
#> ----------------------------------
#> no    | 3.84 | 0.21 | [3.44, 4.28]
#> yes   | 1.24 | 0.19 | [0.92, 1.67]

head(estimate_response(model))
#>        cover mined site Predicted     CI_low   CI_high
#> 1 -1.4423172   yes VF-1 0.2655766 0.14036485 0.3907883
#> 2  0.2984104   yes VF-2 0.1856924 0.09956216 0.2718226
#> 3  0.3978806   yes VF-3 0.1716182 0.08658513 0.2566512
#> 4 -0.4476157    no  R-1 2.4140644 1.83428916 2.9938396
#> 5  0.5968209    no  R-2 2.1956535 1.76590990 2.6253971
#> 6  1.3428470    no  R-3 1.9185412 1.44017437 2.3969079

estimate_smooth(model)
#> No numeric variable was specified for smooth analysis. Selecting `smooth = "cover"`.
#> Part | Start |  End | Size |     Trend | Linearity
#> --------------------------------------------------
#> 1    | -1.59 | 1.89 | 100% | -8.36e-04 |      0.99

estimate_contrasts(model)
#> Level1 | Level2 | Difference |   SE |       95% CI |    t |  df |      p | Difference (std.)
#> --------------------------------------------------------------------------------------------
#> no     |    yes |       1.13 | 0.17 | [0.81, 1.46] | 6.80 | 637 | < .001 |              1.13

estimate_link(model)
#> Error in c1[rownames(c2), colnames(c2), drop = FALSE]: subscript out of bounds
```

<sup>Created on 2020-06-25 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by strengejacke** [comment]
I think this can be merged, we can fix minor issues later, I'd say...

**Comment by strengejacke** [comment]
@DominiqueMakowski will you find time to review this the next days?

**Comment by DominiqueMakowski** [comment]
I'll merge first and review asap!

**Comment by DominiqueMakowski** [comment]
> I'm not sure why you need to set random = FALSE here:

I'm not sure either, I think that the current management of the random argument and its transformation from a simple bool to the appropriate required arg in the predict functions (which can be NULL, NA or some formulas, never really managed to remember how it works exactly) could be reviewed/double checked

> All except estimate_link() should work. I can't figure out why estimate_link() fails for my test model?

I'll check now

**Comment by DominiqueMakowski** [comment]
Yes, the error in estimate_link boils down to `re.form must equal NULL, NA, or ~0`. @strengejacke Do you know if it's possible to predict based on new data without the random effects? Or do we need to include the column corresponding to the RE and fill it with like the reference level or something?

**Comment by DominiqueMakowski** [comment]
```r
  # Deal with random
  if (insight::model_info(model)$is_mixed & random) {
    if (!insight::find_random(model, flatten = TRUE) %in% names(data)) {
      warning("Could not find random effects in data. Will turn `random` to FALSE.")
      random <- FALSE
    }
  }
  if (isTRUE(random)) {
    re.form <- NULL
  } else {
    re.form <- NA
  }
```

This is the current logic used in `.estimate_response_init` but I reckon it's far from optimal. 

Essentially my idea is that the default for `estimate_response()` is that it predicts based on its own data (or on new "real" data), which include the random effects. Thus, it can take it into account. However, in `estimate_link()` the idea is that by default it makes predictions based on a visualization matrix in which we don't want to fix to a given random factor level if possible (but rather obtain estimates averaged over them).

**Comment by strengejacke** [comment]
`predict.glmmTMB()` reuires all variables in `newdata` (and thus, `visualization_matrix.glmmTMB()` should include random terms, setting them to `NA` - this should work now).

**Comment by strengejacke** [comment]
> However, in estimate_link() the idea is that by default it makes predictions based on a visualization matrix in which we don't want to fix to a given random factor level if possible (but rather obtain estimates averaged over them).

And this _should_ work when including the variables from the random effects, setting them to NA. Not sure why this fails here...

**Comment by strengejacke** [comment]
I think I know where the problem is: `visualization_matrix()` does actually not return a _matrix_...

`estimate_link()` calls `.estimate_response_init()`, which call `visualisation_matrix(model, random = FALSE, length = 25, preserve_range = TRUE, reference = insight::get_data(model))`. The latter returns no "symmetric" data frame, which would be expected from a "real" matrix. This fails for `predict()`.

Here is what `visualization_matrix()` returns.

```r
         cover mined site
1  -1.59152240   yes   NA
2  -1.45226419   yes   NA
3  -1.31300598   yes   NA
4  -1.17374777   yes   NA
5  -1.03448956   yes   NA
6  -0.89523135   yes   NA
7  -0.75597314   yes   NA
8  -0.61671493   yes   NA
9  -0.47745672   yes   NA
10 -0.33819851   yes   NA
11 -0.19894030   yes   NA
12 -0.05968209   yes   NA
13  0.07957612   yes   NA
14  0.21883433   yes   NA
15  0.35809254   yes   NA
16 -0.89523135    no   NA
17 -0.75597314    no   NA
18 -0.61671493    no   NA
19 -0.47745672    no   NA
20 -0.33819851    no   NA
21 -0.19894030    no   NA
22 -0.05968209    no   NA
23  0.07957612    no   NA
24  0.21883433    no   NA
25  0.35809254    no   NA
26  0.49735075    no   NA
27  0.63660896    no   NA
28  0.77586717    no   NA
29  0.91512538    no   NA
30  1.05438359    no   NA
31  1.19364180    no   NA
32  1.33290001    no   NA
33  1.47215822    no   NA
34  1.61141643    no   NA
35  1.75067464    no   NA
36  1.88993285    no   NA
```

Here is a solution that works, using `expand.grid()`.

``` r
require("modelbased")
require("glmmTMB")
data(Salamanders)
model <- glmmTMB(
  count ~ cover + mined + (1 | site),
  ziformula = ~ cover + mined,
  family = truncated_poisson,
  data = Salamanders
)

nd <- visualisation_matrix(model, random = FALSE, length = 26, preserve_range = TRUE, reference = insight::get_data(model))
predict(model, newdata = nd, re.form = NA)
#> Error in c1[rownames(c2), colnames(c2), drop = FALSE]: subscript out of bounds

nd <- expand.grid(cover = seq(-2, 2, .2), mined = c("yes", "no"), site = NA)
predict(model, newdata = nd, re.form = NA)
#>  [1]  0.61128095  0.57143738  0.53159382  0.49175025  0.45190668  0.41206311
#>  [7]  0.37221954  0.33237597  0.29253240  0.25268883  0.21284527  0.17300170
#> [13]  0.13315813  0.09331456  0.05347099  0.01362742 -0.02621615 -0.06605972
#> [19] -0.10590329 -0.14574685 -0.18559042  1.74373958  1.70389601  1.66405244
#> [25]  1.62420888  1.58436531  1.54452174  1.50467817  1.46483460  1.42499103
#> [31]  1.38514746  1.34530389  1.30546033  1.26561676  1.22577319  1.18592962
#> [37]  1.14608605  1.10624248  1.06639891  1.02655534  0.98671177  0.94686821
```

**Comment by DominiqueMakowski** [comment]
That's so weird! It fails with:

```r
d <- visualisation_matrix(model, random = FALSE, length = 4, preserve_range = FALSE, reference = insight::get_data(model))
> d
       cover mined site
1 -1.5915224   yes   NA
2 -0.4310373   yes   NA
3  0.7294478   yes   NA
4  1.8899329   yes   NA
5 -1.5915224    no   NA
6 -0.4310373    no   NA
7  0.7294478    no   NA
8  1.8899329    no   NA
```

But works with:

```r
> expand.grid(cover = seq(-2, 2, 1), mined = c("yes", "no"), site = NA)
   cover mined site
1     -2   yes   NA
2     -1   yes   NA
3      0   yes   NA
4      1   yes   NA
5      2   yes   NA
6     -2    no   NA
7     -1    no   NA
8      0    no   NA
9      1    no   NA
10     2    no   NA
```

I don't understand why! the class is the same, the types of the variables are the same, the range of the former is smaller than of the latter...

**Comment by DominiqueMakowski** [comment]
Let's merge this and fix the estimate_link once we have more info for https://github.com/glmmTMB/glmmTMB/issues/602 ☺️ thanks a lot Daniel for this!

**Comment by mattansb** [comment]
> predict(model, newdata = nd, re.form = NA)
#> Error in c1[rownames(c2), colnames(c2), drop = FALSE]: subscript out of bounds

``` r
library(glmmTMB)

data(Salamanders)

model <- glmmTMB(
  count ~ cover + mined + (1 | site),
  ziformula = ~ cover + mined,
  family = truncated_poisson,
  data = Salamanders
)

d <- structure(list(
  cover = c(-1.591522402, -0.431037317, 0.729447768, 
            1.889932853, -1.591522402, -0.431037317, 0.729447768, 1.889932853),
  mined = structure(c(2L, 2L, 2L, 2L, 1L, 1L, 1L, 1L), .Label = c("no","yes"), class = "factor"), 
  site = c(NA, NA, NA, NA, NA, NA, NA, NA)
), class = "data.frame", row.names = c(NA, -8L))

predict(model, newdata = d, re.form = NA)
#> Error in c1[rownames(c2), colnames(c2), drop = FALSE]: subscript out of bounds
```

But…

``` r
levels(d$mined) <- c("yes", "no") # so same as:

levels(insight::get_data(model)$mined)
#> [1] "yes" "no"

predict(model, newdata = d, re.form = NA)
#> [1]  1.66236356  1.43117422  1.19998488  0.96879555  0.52990493  0.29871559
#> [7]  0.06752625 -0.16366308
```

<sup>Created on 2020-06-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


Seems that `predict.glmmTMB` is sensitive to the factor's label order...

**Comment by strengejacke** [comment]
Might be related to https://github.com/easystats/modelbased/issues/68

**Comment by DominiqueMakowski** [comment]
Indeed!

**Comment by mattansb** [comment]
image

**Pull request merged**